### PR TITLE
Add scatter_add

### PR DESCRIPTION
**Pull request by yuyu2172**
Resolve https://github.com/chainer/chainer/issues/3172

### CODE REVIEW

File: chainer/functions/array/scatter_add.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    def forward(self, xs):
+        a = xs[0]
+        b = xs[1]
+        y = a.copy()
+        xp = chainer.cuda.get_array_module(a)
+        if y[self.slices].shape != b.shape:
+            raise ValueError(
+                'Chainer does not support automatic broadcasting '
+                'of variables.')
+        if xp == numpy:
```
Comment (niboshi) [inline chainer/functions/array/scatter_add.py]: Please use `is` operator.

File: chainer/functions/array/scatter_add.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        It only supports types that are supported by CUDA's atomicAdd when
+        an integer array is included in ``slices``.
+        The supported types are ``numpy.float32``, ``numpy.int32``,
+        ``numpy.uint32``, ``numpy.uint64`` and ``numpy.ulonglong``.
+
+    .. note::
+
+        It does not support ``slices`` that contains multiple boolean arrays.
+
+    .. seealso:: :func:`numpy.add.at`.
```
Comment (niboshi) [inline chainer/functions/array/scatter_add.py]: How about listing `cupy.scatter_add` as well?

**Comment by yuyu2172** [comment]
Thank you for the comments. I updated the code.

File: chainer/functions/__init__.py [hunk 3]
```diff
@@ -68,6 +68,8 @@
 from chainer.functions.array.resize_images import ResizeImages  # NOQA
 from chainer.functions.array.rollaxis import rollaxis  # NOQA
 from chainer.functions.array.rollaxis import Rollaxis  # NOQA
+from chainer.functions.array.scatter_add import scatter_add  # NOQA
+from chainer.functions.array.scatter_add import ScatterAdd  # NOQA
```
Comment (niboshi) [inline chainer/functions/__init__.py]: Could you remove `Function` class alias (capital one) for now?
We're currently reconsidering whether class aliases should be made public.

Comment (yuyu2172) [reply chainer/functions/__init__.py]: OK

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM, Thanks!

**Pull request merged**